src/nbl/video/COpenGLCommandBuffer.cpp

```diff
@@ -0,0 +1,743 @@
+                    gl->extGlCopyNamedBufferSubData(readb, writeb, cp.srcOffset, cp.dstOffset, cp.size);
+                }
+            }
+            break;
+            case ECT_COPY_IMAGE:
+            {
+                auto& c = cmd.get<ECT_COPY_IMAGE>();
+                // TODO flush some state? -- not needed i think
+                IGPUImage* dstImage = c.dstImage.get();
+                IGPUImage* srcImage = c.srcImage.get();
+                if (!dstImage->validateCopies(c.regions, c.regions + c.regionCount, srcImage))
+                    return;
+
+                auto src = static_cast<COpenGLImage*>(srcImage);
+                auto dst = static_cast<COpenGLImage*>(dstImage);
+                IGPUImage::E_TYPE srcType = srcImage->getCreationParameters().type;
+                IGPUImage::E_TYPE dstType = dstImage->getCreationParameters().type;
+                constexpr GLenum type2Target[3u] = { GL_TEXTURE_1D_ARRAY,GL_TEXTURE_2D_ARRAY,GL_TEXTURE_3D };
+                for (auto it = c.regions; it != c.regions + c.regionCount; it++)
+                {
+                    gl->extGlCopyImageSubData(src->getOpenGLName(), type2Target[srcType], it->srcSubresource.mipLevel,
+                        it->srcOffset.x, srcType == IGPUImage::ET_1D ? it->srcSubresource.baseArrayLayer : it->srcOffset.y, srcType == IGPUImage::ET_2D ? it->srcSubresource.baseArrayLayer : it->srcOffset.z,
+                        dst->getOpenGLName(), type2Target[dstType], it->dstSubresource.mipLevel,
+                        it->dstOffset.x, dstType == IGPUImage::ET_1D ? it->dstSubresource.baseArrayLayer : it->dstOffset.y, dstType == IGPUImage::ET_2D ? it->dstSubresource.baseArrayLayer : it->dstOffset.z,
+                        it->extent.width, dstType == IGPUImage::ET_1D ? it->dstSubresource.layerCount : it->extent.height, dstType == IGPUImage::ET_2D ? it->dstSubresource.layerCount : it->extent.depth);
+                }
+            }
+            break;
+            case ECT_COPY_BUFFER_TO_IMAGE:
+            {
+                auto& c = cmd.get<ECT_COPY_BUFFER_TO_IMAGE>();
+
+                copyBufferToImage(c, gl, ctxlocal, ctxid);
+            }
+            break;
+            case ECT_COPY_IMAGE_TO_BUFFER:
+            {
+                auto& c = cmd.get<ECT_COPY_IMAGE_TO_BUFFER>();
+
+                copyImageToBuffer(c, gl, ctxlocal, ctxid);
+            }
+            break;
+            case ECT_BLIT_IMAGE:
+            {
+                auto& c = cmd.get<ECT_BLIT_IMAGE>();
+
+                // no way to do it easily in GL
+                // idea: have one 1 fbo per function table (effectively per context) just to do blits and resolves
+            }
+            break;
+            case ECT_RESOLVE_IMAGE:
+            {
+                auto& c = cmd.get<ECT_RESOLVE_IMAGE>();
+
+                // no way to do it easily in GL
+                // idea: have one 1 fbo per function table (effectively per context) just to do blits and resolves
```

Comment:
construct a hash on the go from the input and output images, create FBOs in the LRU Cache of FBOs

Reply:
have a LRU cache of FBOs yet?

would be useful for emulating glGetTexSubImage on ES 3.1

Reply:
yes, we already use it for blit and resolve